torch/jit/__init__.py

```diff
@@ -1731,6 +1731,31 @@ def __getstate__(self):
                 "Mixed serialization of script and non-script modules is not supported. " +
                 "For purely script modules use my_script_module.save(<filename>) instead.")
 
+        # Python magic methods do lookups on an object's class type, instead of looking up
+        # the desugared attribute. In order to continue to expose the magic methods
+        # of builtin-containers (ModuleList, Sequential, ModuleDict) to python we
+        # define magic methods here as a shim to the correct attribute.
+        def forward_magic_method(self, method_name, *args, **kwargs):
```

Comment:
It's a little unfortunate that we are littering these magic methods for `dir()` and `inspect` to pick up. One alternative is for `create_script_module` to generate a new type that inherits from `RecursiveScriptModule` and define all `@_copy_to_script` methods on it. Can we try that?

Reply:
I think it's preferable to separate mechanism from policy, and have the `create_script_module` only know about the mechanism. It's better that the decorators visible I think; if we defined new generic container/sequence classes, or change the magic methods we define on existing ones, we shouldn't expect implementors to know that they need to go to `_recursive` to change. We can consult with the NY team maybe?

As for whether we will expose this to the user—I don't see anything wrong with making `_copy_to_script` public (it's useful and non-duplicative) but we should be careful about increasing the API surface area so I think in this PR we keep it private.

Reply:
Separately - what exactly were you envisioning with the `generate a new type` comment ?

Reply:
> I also think it's best to avoid the pattern of inserting code that only has semantic meaning to the JIT team in the rest of pytorch when we can; in past we've done so e.g. in LSTM with __overload__ and @ignore and it hasn't been great.

I think this is a judgment call. `__overload__` and `@ignore` complicate up the code in parts that need to be readable, this decorator does not. I think it's fine in this case.

As for the "generate a new type" thing, @zdevito convinced me that it's a bad idea. I think we can just use `forward_magic_method` but we need to fix the implementation so that `__dir__` works. Also, it would be easier I think if `forward_magic_method` were called in a loop over a list of names, like we do for copying methods from `RecursiveScriptModule` to `ScriptModule`.

Reply:
I don't know how the "loop over a list of names" would work. The method needs to be defined on the class, so copying these as attributes will not work.

Reply:
Okay, updated to use wrapper. adding pytorch eager ppl to review. cc @zou3519 @gchanan